client-v2/src/test/java/com/clickhouse/client/query/QueryTests.java

```diff
@@ -65,65 +85,368 @@ public void testSimpleSelectTableFormat() {
+
+        Iterator<Map<String, Object>> dataIterator = data.iterator();
         try {
-            RowBinaryReader reader = new RowBinaryReader(response.get().getInputStream());
-            reader.readBatch(100, records::add, System.out::println);
-        } catch (InterruptedException | ExecutionException e) {
-            Assert.fail("failed to get response", e);
+            while (dataIterator.hasNext()) {
+                Map<String, Object> expectedRecord = dataIterator.next();
+                reader.next();
+                reader.copyRecord(record);
+                Assert.assertEquals(record, expectedRecord);
+            }
+        } catch (IOException e) {
+            Assert.fail("failed to read response", e);
         }
     }
 
-    private void prepareDataSet() {
-        try (
-            ClickHouseClient client = ClickHouseClient.builder().config(new ClickHouseConfig())
-                    .nodeSelector(ClickHouseNodeSelector.of(ClickHouseProtocol.HTTP))
-                    .build()) {
+    private static ClickHouseBinaryFormatReader createBinaryFormatReader(QueryResponse response, QuerySettings settings,
+                                                                         TableSchema schema) {
+        ClickHouseBinaryFormatReader reader = null;
+        switch (response.getFormat()) {
+            case Native:
+                reader = new NativeFormatReader(response.getInputStream(), settings);
+                break;
+            case RowBinaryWithNamesAndTypes:
+                reader = new RowBinaryWithNamesAndTypesFormatReader(response.getInputStream(), settings);
+                break;
+            case RowBinaryWithNames:
+                reader = new RowBinaryWithNamesFormatReader(response.getInputStream(), settings, schema);
+                break;
+            case RowBinary:
+                reader = new RowBinaryFormatReader(response.getInputStream(), settings, schema);
+                break;
+            default:
+                Assert.fail("unsupported format: " + response.getFormat().name());
+        }
+        return reader;
+    }
+
+    @Test(groups = {"integration"})
+    public void testBinaryStreamReader() throws Exception {
+        final String table = "dynamic_schema_test_table";
+        List<Map<String, Object>> data = prepareDataSet(table, DATASET_COLUMNS,
+                DATASET_VALUE_GENERATORS, 10);
+        QuerySettings settings = new QuerySettings().setFormat(ClickHouseFormat.RowBinaryWithNamesAndTypes.name());
+        Future<QueryResponse> response = client.query("SELECT col1, col3, hostname() as host FROM " + table, null, settings);
+        QueryResponse queryResponse = response.get();
+
+        TableSchema schema = new TableSchema();
+        schema.addColumn("col1", "UInt32");
+        schema.addColumn("col3", "String");
+        schema.addColumn("host", "String");
+        ClickHouseBinaryFormatReader reader = createBinaryFormatReader(queryResponse, settings, schema);

```

Comment:
I found this API a little bit overcomplicated. Didn't we agree to have a records method that you can pull the data from it ?

Reply:
Using records means adding another layer and memory overhead, I think. So I've done 2 ways of getting data from result 
- Iterator style and getters (what I thought was the idea) 
- com.clickhouse.client.api.data_formats.ClickHouseBinaryFormatReader#copyRecord method to copy each record to a destination map to let user process it. 

What do you think is working not well?